### ISSUE TITLE
Non-constant default values

### ISSUE THREAD

**Issue by sdebruyn** [rrousselGit/freezed#64]
Thank you for the 0.7.0 update, the default values made my code a little bit simpler :)

One thing that I still have to work around are non-constant default values. My type has a `String` property called `id` that - if left empty - should have a random UUID value by default. I now work around this by adding a second factory method that redirects to the generated one without the `id` parameter.

It would be nice if the `@Default` annotation could also take in a lambda to generate the default value.

**Comment by rrousselGit** [comment]
Dart only supports constant default values.

Are you looking for `@late` instead?

**Comment by rrousselGit** [comment]
Ah, I see what you're speaking of.
But the syntax using `@Default` would be horrible as we can't use `() => something` inside decorators.

What about such syntax?

@freezed
abstract class SuperLate with _$SuperLate {
  factory SuperLate([int value]) = _SuperLate;

  @late
  int get value => value ?? Random().nextInt(9999);
}

**Comment by sdebruyn** [comment]
Yes, that would be great! The `late` feature isn't enough because it doesn't allow the user to supply a custom value.

**Comment by rrousselGit** [comment]
We'll probably need an alternate syntax for unions too:

```dart
@freezed
abstract class SuperLate with _$SuperLate {
  factory SuperLate([@MyDefault() int value]) = _SuperLate;
  factory SuperLate.emoty() = _SuperLate;
}

class MyDefault implements Default<T> {
  const MyDefault();

  T get defaultValue => Random().nextInt(9999);
}
```

Kinda verbose though.

**Comment by bounty1342** [comment]
It matched Json converter template (in a sense that is good).
https://github.com/dart-lang/json_serializable/blob/master/example/lib/json_converter_example.dart
Depending if this is an instance of Default() or JsonConverter(), will handle generation.

If this gets done, does this mean, we could have default non literals also ?

**Comment by tbm98** [comment]
Hi @rrousselGit.
I write as 

> Ah, I see what you're speaking of.
> But the syntax using `@Default` would be horrible as we can't use `() => something` inside decorators.
> 
> What about such syntax?
> 
> ```dart
> @freezed
> abstract class SuperLate with _$SuperLate {
>   factory SuperLate([int value]) = _SuperLate;
> 
>   @late
>   @override
>   int get value => super.value ?? Random().nextInt(9999);
> }
> ```

but get error
<img width="788" alt="Screen Shot 2020-06-25 at 8 32 59 PM" src="https://user-images.githubusercontent.com/52562340/85729754-4524cf00-b723-11ea-85df-790217252c4c.png">

**Comment by rrousselGit** [comment]
This is not implemented. Just a thought

**Comment by Aqluse** [comment]
@rrousselGit Could this thought came to reality? I got into the same problem when I want to set empty list to a some field, but I can't get it generified in the `@Default `, and generator come broken when it try to implement it with writing block body, returning full-signatured constructor call with defaults written in this call. Like this:
```
factory RepositoryListDataModel({
    @Default(false) bool fetchCompleted,
    String query,
    List<T> data
  }) = {
    return _RepositoryListDataModel<T>(query: query, data: data ?? <T>[]);
  }
```

**Comment by aloisdeniel** [comment]
One way to implement this would be to allow functions as `Default`'s argument :

```dart
Other _defaultOther() => Other();

@freezed
abstract class Example with _$Example {
  factory Example(@Default(_defaultOther) Other value) = _Example;
}
```

**Comment by aloisdeniel** [comment]
Yeah, probably the most common scenario.

But it may solve issues with json serializable :

For example :

```dart
@freezed
abstract class Other with _$Other {
  const factory Other() = _Other;

  factory Example.fromJson(Map<String, dynamic> json) => _$ExampleFromJson(json);
}

@freezed
abstract class Example with _$Example {
  const factory Example(@Default(const Other()) Other value) = _Example;

  factory Example.fromJson(Map<String, dynamic> json) => _$ExampleFromJson(json);
}
```

That produces this error :

> Error with `@JsonKey` on `value`. `defaultValue` is `_$_Other`, it must be a literal.

**Comment by aloisdeniel** [comment]
Another common use cas that doesn't need other properties :

```dart
Example({
    DateTime timestamp,
  }) : timestamp = timestamp ?? DateTime.now().add(Duration(days: -30));
```

**Comment by hcbpassos** [comment]
@rrousselGit 

> The problem is, people will most likely want to initialize some parameters based on other parameters

I understand your concern, but I feel like it's a way to eagerly address future problems, since there hasn't been such feature request until now, as far as I could evaluate. I must add that I cannot really find a good use of such feature. Default values that depend on different parameters are something other than just default values. I cannot find a good reason to make it easy embracing bad practices.

**Comment by hcbpassos** [comment]
@aloisdeniel If we got to that point, why not creating a dummy private constructor and a second one with body?

```dart
const factory Example._(DateTime timestamp) = _Example;

factory Example({
  DateTime timestamp,
}) => Example._(timestamp ?? DateTime.now().add(Duration(days: -30)));
```

**Comment by rrousselGit** [comment]
> I understand your concern, but I feel like it's a way to eagerly address future problems, since there hasn't been such feature request until now, as far as I could evaluate.
> 
> I must add that I cannot really find a good use of such functionality. Default values that depend on different parameters are something other than just default values. I cannot find a good reason to make it easy embracing bad practices.

People have requested a few times a way to generate an ID

Something like:

```dart
Constructor({
  String? id,
}): id = id ?? uuid.v4();
```

**Comment by rrousselGit** [comment]
> Notice that the parameter is not optional in the dummy constructor. This is a very important detail. By making it optional, freezed would not generate assert(timestamp != null).

Not if you mark the parameter as `required`.

**Comment by aloisdeniel** [comment]
@hcbpassos Yes absolutely, it is another solution, but I wanted to keep the `const` behaviour. :)

**Comment by Valentin-Seehausen** [comment]
Any progress on this? I would like to see this functionality because in a truly null safe app I need default values when I read from the database. It would be awesome to have these default constructors.

**Comment by hcbpassos** [comment]
@aloisdeniel 
> Yes absolutely, it is another valid solution, but I wanted to keep the const behaviour. :)

I'm not sure I understand. Do you want a const constructor with non-const default values?

**Comment by elianmortega** [comment]
I have kind of the same issue, don't focus on the logic i was just trying to test forms with riverpod state notifier and freeezed,

but I'm not able to continue because DateTime doesn't have a const constructor so I'm thinking is there a possibility to do it or i just need to make a workaround for it?

```dart
@freezed
class LoginState with _$LoginState {
  factory LoginState({
    @Default(Email.pure()) Email email,
    @Default(Password.pure()) Password password,
    @Default(DateTime()) Date date,
    @Default(FormzStatus.pure) FormzStatus status,
  }) = _LoginState;
}

```

**Comment by Valentin-Seehausen** [comment]
Is there another proposed solution to use another freezed class as a default value? Like a default value for a subclass?

**Comment by SalahAdDin** [comment]
> Yeah, probably the most common scenario.
> 
> But it may solve issues with json serializable :
> 
> For example :
> 
> ```dart
> @freezed
> abstract class Other with _$Other {
>   const factory Other() = _Other;
> 
>   factory Example.fromJson(Map<String, dynamic> json) => _$ExampleFromJson(json);
> }
> 
> @freezed
> abstract class Example with _$Example {
>   const factory Example(@Default(const Other()) Other value) = _Example;
> 
>   factory Example.fromJson(Map<String, dynamic> json) => _$ExampleFromJson(json);
> }
> ```
> 
> That produces this error :
> 
> > Error with `@JsonKey` on `value`. `defaultValue` is `_$_Other`, it must be a literal.
> 
> I'm clueless on how to achieve this right now.

I have the same problem when i want to put an model default value:
```dart
@freezed
class User with _$User {
  const factory User(
      {required String id,
      required String email,
      required String username,
      @Default(Profile.empty) Profile profile,}) = _User;

  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
  static const empty = User(
      id: '',
      email: 'name@dummy.com',
      username: 'DummyName',);
}
```

Where `Profile` is the model we need:
```shell
[SEVERE] json_serializable:json_serializable on lib/features/user/domain/user.entity.dart:

Error with `@JsonKey` on `profile`. `defaultValue` is `_$_Profile`, it must be a literal.
package:thesis_cancer/features/user/domain/user.entity.freezed.dart:297:17
    ╷
297 │   final Profile profile;
    │                 ^^^^^^^
    ╵
[INFO] Running build completed, took 4.6s

[INFO] Caching finalized dependency graph...
[INFO] Caching finalized dependency graph completed, took 80ms

[SEVERE] Failed after 4.7s
pub finished with exit code 1
```

**Comment by rodion-m** [comment]
What about cases with default list value? Like this:
```dart
class Account with _$Account {
  factory Account({
    String name,
    @Default([]) List<String> photos, //<--- this is not working
  }) = _Account;
}
```
I just want to init empty list and don't want to use null. How to do that?

**Comment by SunlightBro** [comment]
@rodion-m you shouldn't use implicit-cast but instead properly type your empty list:
```dart
 @Default(<String>[]) List<String> photos,
```

**Comment by rodion-m** [comment]
> @rodion-m you can't use implicit-cast for your default value, instead give your empty list a proper type:
> 
> ```dart
>  @Default(<String>[]) List<String> photos,
> ```

Thank you, it works!

**Comment by rrousselGit** [comment]
> ```dart
> @Default(Set<String>.unmodifiable(<String>[])) Set<String> photos;
> ```
> 
> any idea how i can map the above code on a Set? with this code, i do get _The constructor being called isn't a const constructor..._

Instead do:

```dart
@Default(<String>{})
```

**Comment by exaby73** [comment]
@rrousselGit Hey. Do you have plans in the near future to solve this? It's been a few months now

**Comment by rrousselGit** [comment]
No. Feel free to make a pull request

**Comment by exaby73** [comment]
@rrousselGit As soon as I am able to wrap my head around the Dart build system, I'd love to. BTW your efforts, and that of the community, is well appreciated. I'd love to donate to this project to keep it alive even if all I can afford is a few bucks :D

**Comment by henk-lim** [comment]
FYI. For those looking for a workaround until this is implemented. You can keep the DateTime field nullable. Or...

A way around this problem is to use a JsonConverter for each DateTime field as follow:

Given a field: `    @DateTimeJsonConverter() required DateTime birthdate,`

```

class DateTimeJsonConverter implements JsonConverter<DateTime, String?> {
  const DateTimeJsonConverter();

  @override
  DateTime fromJson(String? json) => json == null ? DateTimeX.startOfTime : DateTime.parse(json);

  @override
  String toJson(DateTime object) => object.toIso8601String();
}
```

This will always default to a given value when converting from Json. However, the required keyword will essentially mean that the model can no longer be treated as a constant and it must be evaluated at run time. Which means that it cannot be used in itself as a default value in another model. That is to say that if a model is dependent on a dateTime object, or an object that depends on a dateTime object, and so on, then it cannot be used as a const literal going forward.

While you cannot use the object as a literal, you can still have an 'initial value' at runtime (which can be used for comparison). Example:

`  static User get empty => User(birthdate: DateTimeX.startOfTime);`

**Comment by Giuspepe** [comment]
> Thank you for the 0.7.0 update, the default values made my code a little bit simpler :)
> 
> One thing that I still have to work around are non-constant default values. My type has a `String` property called `id` that - if left empty - should have a random UUID value by default. I now work around this by adding a second factory method that redirects to the generated one without the `id` parameter.
> 
> It would be nice if the `@Default` annotation could also take in a lambda to generate the default value.

Hey @sdebruyn, could you please provide a code example for your workaround? I have a similar use case where I need a random UUID as default value but didn't find a workaround with freezed yet.

**Comment by SalahAdDin** [comment]
> > Thank you for the 0.7.0 update, the default values made my code a little bit simpler :)
> > One thing that I still have to work around are non-constant default values. My type has a `String` property called `id` that - if left empty - should have a random UUID value by default. I now work around this by adding a second factory method that redirects to the generated one without the `id` parameter.
> > It would be nice if the `@Default` annotation could also take in a lambda to generate the default value.
> 
> Hey @sdebruyn, could you please provide a code example for your workaround? I have a similar use case where I need a random UUID as default value but didn't find a workaround with freezed yet.

I'm interesting on it too.

**Comment by thipokch** [comment]
@Giuspepe @SalahAdDin here's the workaround
``` dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:uuid/uuid.dart';

class DefaultUuid implements JsonConverter<String, String> {
  const DefaultUuid();

  @override
  String fromJson(String json) {
    if (json is String && json.isNotEmpty) {
      return json;
    }

    return const Uuid().v4();
  }

  @override
  String toJson(String data) => data ?? const Uuid().v4();
}
```

Then you would use this as `@Default(DefaultUuid())`

**Comment by tobytraylor** [comment]
The way that this was solved in [build value](https://github.com/google/built_value.dart) by David Morgan was to identify a static function with a attribute that would be called to perform the defaulting.

`abstract class MyValue {
  @BuiltValueHook(initializeBuilder: true)
  static void _setDefaults(MyValueBuilder b) =>
      b
        ..name = 'defaultName'
        ..count = 0;`

[built value issue / discussion](https://github.com/google/built_value.dart/issues/212)

**Comment by Shiba-Kar** [comment]
how to assign a default datetime 
```dart
@freezed
class Client with _$Client {
  factory Client({
    required User user,
    @Default(DateTime.now()) DateTime addedDate, //The constructor being called isn't a const constructor.
Try removing 'const' from the constructor invocation.
  }) = _Client;
@late
@override
  factory Client.fromJson(Map<String, dynamic> json) => _$ClientFromJson(json);
}

````

**Comment by feduke-nukem** [comment]
> 

Does that suppose to work?

got this message ```is non-nullable but is neither required nor marked with @Default```

**Comment by tony-asu** [comment]
I got it to work by making my own inherited Default

```dart
class DefaultDateTimeNow extends Default {
  const DefaultDateTimeNow() : super('');
  @override
  Object? get defaultValue => DateTime.now();
}
```

and to give a datetime a default value of now.


```dart
@freezed
class Client with _$Client {
  factory Client({
    @DefaultDateTimeNow() DateTime addedDate,
...
```

**Comment by paramsinghvc** [comment]
Uhh, I miss Typescript! Too much work for union types!

**Comment by RyoheiTomiyama** [comment]
I resolved like this.

```dart
typedef UUID = String;

@freezed
class SendData with _$SendData {
  factory SendData.def({
    required String receiver_id,
    required List<FileData> files,
    required UUID uuid,
  }) = _SendData;

  factory SendData({
    required String receiver_id,
    required List<FileData> files,
  }) {
    return _SendData(
      receiver_id: receiver_id,
      files: files,
      uuid: _generateUuid(),
    );
  }

  static UUID _generateUuid() {
    final uuid = Uuid();
    return uuid.v4();
  }
}

```

```dart
final sendData = SendData(receiver_id: receiver_id, files: files);
```

**Comment by normidar** [comment]
Any progress on this?

https://github.com/rrousselGit/freezed/issues/64#issuecomment-1555921659
this is diffcalt difficult when I hava lot of field in my class.

**Comment by fidelQabit** [comment]
I have problem with this and DateTime, i have to make that fields nullable by default instead of DateTime.now() lol:

```dart
DateTime? metricsDateFrom,
DateTime? metricsDateTo,
```
